### DISCUSSION TITLE
How to stop ignoring _ide_helper.php?

### DISCUSSION THREAD

**Discussion by Chekote** [composer-unused/composer-unused#660] [Q&A] [answered]
Howdy!

First, thanks for creating this very helpful tool!

I am running this on my Laravel app but it is flagging many packages as unused because they are used via the Facades and not direct references to the symbols in the package. We use `barryvdh/laravel-ide-helper` to help tooling understand how to resolve the Facades to the appropriate package, but it does not seem to help with composer-unused, as it still reports the packages as unused even with the `ide_helper.php` file in place.

I did some experiments with one of the packages `sentry/sentry-laravel`, which is claimed to be unused, but I know for a fact it is not. My experiment consistent of me putting the line `echo \Sentry\Laravel\Facade::class;` at the beginning of a few files to see if I can gain some better understanding of why composer-unused is claiming the package is unused. These are my results:

| Adding to file | result |
|-|-|
| /_ide_helper.php | still shows as unused |
| /server.php | still shows as unused |
| /app/Http/Controllers/MyController.php| shows as used |

I don't see any way to configure the tool to include the files that are not having the expected effect, and I am not having luck finding anything in the source code for the project that would explain why they are ignored. Can someone please help me understand what is happening here and how to overcome this problem?

**Comment by icanhazstring** [comment]
Hi. Thanks for reaching out.
Typically, only files from the composer autoload directives are scanned for symbols and usage.

If you want to point composer-unused to additional files, you can do that using the configuration file.

**Reply by Chekote** [reply]
I tried using the config to add additional files, but it did not work the way I expected. Maybe I am doing it wrong.

I had initially assumed that I could configure the tool to scan `_ide_helper.php` for usage, but it does not seem possible to do so. For example, I would expect to be able to call something like `$config->setAdditionalFiles([<list-of-file-paths>]);`, which would simply add the files to the scan and dynamically detect which symbols from which packages are used, but nothing like this exists as far as I can tell.

The only thing I was able to find was the following:

`$config->setAdditionalFilesFor('dependency/name', [<list-of-file-paths>]);`

But after experimenting with this, it does not seem to work as I expect. For example, let's take the package `cweagans/composer-patches` which is used by Composer, but is not referenced in the project at all, even in `_ide_helper`. If I create the following config:

```
<?PHP

declare(strict_types=1);

use ComposerUnused\ComposerUnused\Configuration\Configuration;

return static function (Configuration $config): Configuration {
    $config->setAdditionalFilesFor('cweagans/composer-patches', [__DIR__ . '/_ide_helper.php']);

    return $config;
};
```

Then composer-unused said the package is used. So I honestly don't understand what the `setAdditionalFilesFor()` method is supposed to be doing, since by simply referencing a file that exists, it marks the package as used, even if the referenced file does not use it at all.

**Reply by icanhazstring** [reply, marked as answer]
That function adds additional files with used and provided symbols to the given package. Your root project is considered a package, for that your `composer.json` has a `name` attribute.

Pass that value into the first argument of that function. This will tell composer-unused to add found and used symbols to your project. Which in return can then mark required packages to be used.

**Reply by Chekote** [reply]
Awesome. Thank you so much for your help and your prompt responses 🙇
